OptaPy Blog Post (#405)
kiegroup/optaplanner-website: content/blog/2021/10/05/OptimizeYourPython.adoc

> = Optimize your Python

Reviewer: This title implies that we're optimizing Python code.
Instead, we're optimizing planning problem with Python.

Also, it would be good to have keywords in this title, like "AI" and "constraint solver/solving".

A) How about "A new AI constraint solver for Python: OptaPy"
B) "Want to use OptaPlanner in Python? Now you can!" Nah, I wouldn't target the messaging to "OptaPlanner in Python". We want to attrackt new Python users, not migrate the existing Java users (we are still Java first).
C) "A new Python constraint solver is born: OptaPy" (Burr rule: most important keyword first)

-1 for B. I kinda prefer A), but C) is good too. Other proposals welcome :)

Christopher-Chianelli
2021-10-05
:page-interpolate: true
:jbake-type: post

> :jbake-tags: python, integration

Reviewer: Can we add a :jbake-social_media_share_image: ?

How about something like [optaplanner icon + python icon = "OptaPy"]?
We can add it under the first paragraph too. I 'll quickly create it and mail it to you - your call if you want to add it.

> OptaPlanner can now be used in Python using https://pypi.org/project/optapy[OptaPy]!

Author: https://pypi.org/project/optapy works fine for me

> Let's create the school timetabling example in pure Python. Feel free to follow along in the https://mybinder.org/v2/gh/optapy/optapy/main?filepath=index.ipynb[OptaPy Jupyter notebook].

Reviewer: Ok, the first paragraph has a few goals that it needs to attain:

- Convince python people to take a look at this. So ROI. The words "use OptaPlanner in Python" isn't convincing for python users (they don't know optaplanner).
- Assure them that they can use plain Python. "Pure python" is good already, but I think "plain Python" and mentioning pype might be better.
- The binder message is very important, but can be in the 2th paragraph. It's no longer "motivation" (1th para), it's a "call to action" (2th para).
- Assure Java users reading this that we're not migrating to Python. Prove that by mentioning we support Kotlin too. Or the perf difference?
- Explain OptaPy vs OptaPlanner. Explain its essentially a wrapper.

Let me think of a proposal text and suggest it here.

==  Prerequisites

- https://www.python.org/[Install Python 3.9 or later].
- https://adoptopenjdk.net/[Install JDK 11 or later] with the environment variable JAVA_HOME configured to the JDK installation directory.

== Setup

. Create a new Python virtual environment
+
[source,bash]
----
python3 -m venv optapy-env
----

. Activate the Python virtual environment
+
[source,bash]
----
source optapy-env/bin/activate
----

. Install OptaPy using pip
+
[source,bash]
----
python3 -m pip install optapy
----

== School timetabling

In school timetabling, we have a list of lessons that we need to assign to timeslot and rooms.

> Additionally, we want some constraints to be followed:

Reviewer: "to be followed" => passive voice
```suggestion
Additionally, we follow these constraints:
```

> In school timetabling, we have a list of lessons that we need to assign to timeslot and rooms.
> Additionally, we want some constraints to be followed:

Reviewer: timeslot: plural

```suggestion
In school timetabling, we need to assign a list of lessons to timeslots and rooms.
Additionally, there are some constraints:
```

- A room can have at most one lesson at the same time.

- A teacher can teach at most one lesson at the same time.

- A student can attend at most one lesson at the same time.

- A teacher prefers to teach all lessons in the same room.

- A teacher prefers to teach sequential lessons and dislikes gaps between lessons.

- A student dislikes sequential lessons on the same subject.

== Modelling the domain

The Objects used in constraints are known as the domain of the problem. In school timetabling, this consist of lessons, rooms and timeslots.

=== Problem Facts

> Problem facts do not change throughout solving. Rooms and timeslots are examples of problem facts. Let's create the Room class:

Author: IMO, too wordy, and introduce concepts that are not used (and not currently supported in OptaPy)

        self.timeslot = newTimeslot

    @planning_variable(Room, ["roomRange"])
    def getRoom(self):
        return self.room

    def setRoom(self, newRoom):
        self.room = newRoom

    def __str__(self):
        return "Lesson(id=" + str(self.id) + \
                ", timeslot=" + str(self.timeslot) + ", room=" + str(self.room) + \
                ", teacher=" + str(self.teacher) + ", subject=" + str(self.subject) + \
                ", studentGroup=" + str(self.studentGroup) + ")"
----

The `@planning_entity` decorator registers the class as a planning entity, which allows OptaPlanner to assign its planning variables and for it to be used in constraints.
The `@planning_variable(variable_type, [value_range_provider_refs...])` decorator registers a method as the getter of a planning variable.
The getter must be named `get<X>` and the setter must be named `set<X>`.
The first argument, `variable_type`, tells OptaPlanner what type of values OptaPlanner can assign to this planning variable.
The second argument, `value_range_provider_refs`, tells OptaPlanner what value ranges (more info on that later) it takes its values from.

== Constraints

> Constraints tell OptaPlanner the score of the solution. A score represents the quality of a specific solution. The higher the better. OptaPlanner looks for the best solution, which is the solution with the highest score found in the available time. It might be the optimal solution.

Reviewer 2: ```suggestion
Constraints define the score calculation, or the fitness function, of a planning problem. Each solution of a planning problem is graded with a score. A score represents the quality of a specific solution. The higher the score the better. OptaPlanner looks for the best solution, which is the solution with the highest score found in the available time. It might or might not be the optimal solution.
```

                ]) \
            .penalize("Room conflict", HardSoftScore.ONE_HARD)

def teacherConflict(constraintFactory):
    # A teacher can teach at most one lesson at the same time.
    return constraintFactory \
                .fromUniquePair(LessonClass,
                        [
                            Joiners.equal(lambda lesson: lesson.timeslot),
                            Joiners.equal(lambda lesson: lesson.teacher)
                        ]) \
                .penalize("Teacher conflict", HardSoftScore.ONE_HARD)

def studentGroupConflict(constraintFactory):
    # A student can attend at most one lesson at the same time.
    return constraintFactory \
            .fromUniquePair(LessonClass,
                [
                    Joiners.equal(lambda lesson: lesson.timeslot),
                    Joiners.equal(lambda lesson: lesson.studentGroup)
                ]) \
            .penalize("Student group conflict", HardSoftScore.ONE_HARD)
----

The `@constraint_provider` decorator allows OptaPlanner to use a function as a constraint provider.
The function must take exactly one argument; the passed argument is a `ConstraintFactory` used for creating constraints.

> For more information, see https://docs.optaplanner.org/latest/optaplanner-docs/html_single/index.html#constraintStreams[Constraint Streams in the OptaPlanner Docs].

Reviewer 2: ```suggestion
For more information, see https://docs.optaplanner.org/latest/optaplanner-docs/html_single/index.html#constraintStreams[Constraint Streams] in the OptaPlanner documentation.
```

== Gather the domain objects in a planning solution

> A TimeTable wraps all Timeslot, Room, and Lesson instances of a single dataset.

Reviewer 2: ```suggestion
A TimeTable class wraps all Timeslot, Room, and Lesson instances of a single data set.
```

> Furthermore, because it contains all lessons, each with a specific planning variable state, it is a planning solution and it has a score:

Reviewer 2: ```suggestion
Furthermore, because it contains all lessons, each with a specific planning variable state, the TimeTable class is a planning solution and it has a score:
```

                   value_range_provider, planning_score

@planning_solution
class TimeTable:
    def __init__(self, timeslotList=[], roomList=[], lessonList=[], score=None):
        self.timeslotList = timeslotList
        self.roomList = roomList
        self.lessonList = lessonList
        self.score = score

    @problem_fact_collection_property(Timeslot)
    @value_range_provider("timeslotRange")
    def getTimeslotList(self):
        return self.timeslotList

    @problem_fact_collection_property(Room)
    @value_range_provider("roomRange")
    def value_range_provider(self):
        return self.roomList

    @planning_entity_collection_property(Lesson)
    def getLessonList(self):
        return self.lessonList

    @planning_score(HardSoftScore)
    def getScore(self):
        return self.score

    def setScore(self, score):
        self.score = score

    def __str__(self):
        return "TimeTable(timeSlotList=" + str(self.timeslotList) + \
               ",\nroomList=" + str(self.roomList) + ",\nlessonList=" + str(self.lessonList) + \
               ",\nscore=" + (str(self.score.toString()) if self.score is not None else 'None') + ")"
----

> The `@planning_solution` decorator tells OptaPlanner the class `TimeTable` holds the input and output data.

Reviewer 2: ```suggestion
The `@planning_solution` decorator tells OptaPlanner that the class `TimeTable` holds the input and output data.
```

> The `@problem_fact_collection_property(fact_type)` decorator tells OptaPlanner the function provide problem facts.

Reviewer 2: ```suggestion
The `@problem_fact_collection_property(fact_type)` decorator tells OptaPlanner the function that provides problem facts.
```

> The `fact_type` argument tell OptaPlanner what type of fact it provides (for instance, Rooms).

Reviewer 2: ```suggestion
The `fact_type` argument tells OptaPlanner what type of fact it provides (for instance, Rooms).
```

> The `@value_range_provider(range_id)` decorator tell OptaPlanner the function provide a value range, which is used to get possible values of planning variables.

Reviewer 2: ```suggestion
The `@value_range_provider(range_id)` decorator tell OptaPlanner the function that provides a value range, which is used to get possible values of planning variables.
```

Its argument, `range_id` is a string which is used in `@planning_variable` decorators to link the two (for example, `@planning_variable(Room, ['roomRange'])` is linked to `@value_range_provider('roomRange')`.

> The `@planning_entity_collection_property(entity_type)` decorator tells OptaPlanner the function provide planning entities.

Reviewer 2: ```suggestion
The `@planning_entity_collection_property(entity_type)` decorator tells OptaPlanner the function that provides planning entities.
```

The `entity_type` argument tell OptaPlanner what type of entities it provides (for instance, Lessons).

> The `@planning_score(score_type)` decorator tells OptaPlanner the function returns the score.

Reviewer 2: ```suggestion
The `@planning_score(score_type)` decorator tells OptaPlanner the function that returns the score.
```

It must be named `get<X>` and have a corresponding a setter `set<X>`.

> The `score_type` argument tell OptaPlanner what type of score to use (for instance, `HardSoftScore`).

Reviewer 2: ```suggestion
The `score_type` argument tells OptaPlanner what type of score to use (for instance, `HardSoftScore`).
```

|            | 9th grade  | 10th grade |            |
|------------|------------|------------|------------|
| TUE 08:30: | Physics    | Spanish    |            |
|            | M. Curie   | P. Cruz    |            |
|            | 9th grade  | 10th grade |            |
|------------|------------|------------|------------|
| TUE 09:30: | Chemistry  |            | English    |
|            | M. Curie   |            | I. Jones   |
|            | 10th grade |            | 9th grade  |
|------------|------------|------------|------------|
| TUE 10:30: | Physics    | Spanish    |            |
|            | M. Curie   | P. Cruz    |            |
|            | 10th grade | 9th grade  |            |
|------------|------------|------------|------------|
| TUE 13:30: | French     |            | Biology    |
|            | M. Curie   |            | C. Darwin  |
|            | 10th grade |            | 9th grade  |
|------------|------------|------------|------------|
| TUE 14:30: | Chemistry  | Geography  |            |
|            | M. Curie   | C. Darwin  |            |
|            | 9th grade  | 10th grade |            |
|------------|------------|------------|------------|

> ----

Reviewer: I think this output needs to be at the top, to **show** what we're building before we start building it.

Alternatively - even better I think - would be to Inkscape edit the svg next to this png:
https://github.com/kiegroup/optaplanner/blob/main/optaplanner-docs/src/modules/ROOT/images/quickstart/school-timetabling/schoolTimetablingInputOutput.png
and replace the OptaPlanner icon with the words "OptaPy" and use that at the start of this blog post as well as in the optapy.org README. Then the console output can stay at the end.

== Run the application

To run the application, put all the code
into a single Python file and run it:
[source,bash]
----
python3 main.py
----

== Conclusion

> With OptaPy, Python projects can now use OptaPlanner without needing to write any Java code.

Reviewer: ```suggestion
With https://www.optapy.org[OptaPy], Python developers can now use OptaPlanner in plain Python code (no Java coding needed).
```
